lib/IDE/CodeCompletion.cpp

```diff
@@ -1510,12 +1510,19 @@ static Type getReturnTypeFromContext(const DeclContext *DC) {
   if (auto FD = dyn_cast<AbstractFunctionDecl>(DC)) {
     if (FD->hasInterfaceType()) {
       if (auto FT = FD->getInterfaceType()->getAs<FunctionType>()) {
+        if (FD->getDeclContext()->isTypeContext())
+          FT = FT->getResult()->castTo<FunctionType>();
         return FT->getResult();
       }
     }
-  } else if (auto CE = dyn_cast<AbstractClosureExpr>(DC)) {
-    if (CE->getType()) {
-      return CE->getResultType();
+  } else if (auto ACE = dyn_cast<AbstractClosureExpr>(DC)) {
+    if (ACE->getType())
+      return ACE->getResultType();
+    if (auto CE = dyn_cast<ClosureExpr>(ACE)) {
+      if (CE->hasExplicitResultType())
+        return const_cast<ClosureExpr *>(CE)
+            ->getExplicitResultTypeLoc()
+            .getType();
```

Comment:
Why isn't this just set as the result type?

Reply:
I haven't looked into it yet, but probably, if the type checker fails to type check the body of *single expression closure*, type checking for closure also fails.